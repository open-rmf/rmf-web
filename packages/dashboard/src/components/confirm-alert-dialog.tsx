import React from 'react';
import Button from '@material-ui/core/Button';
import Dialog, { DialogProps } from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import Slide from '@material-ui/core/Slide';
import { TransitionProps } from '@material-ui/core/transitions';
import WarningIcon from '@material-ui/icons/Warning';
import { makeStyles } from '@material-ui/core';

const Transition = React.forwardRef(function Transition(
  props: TransitionProps & { children?: React.ReactElement<any, any> },
  ref: React.Ref<unknown>,
) {
  return <Slide direction="up" ref={ref} {...props} />;
});

export interface ConfirmAlertDialogProps extends DialogProps {
  title?: string;
  content?: string;
  showIcon?: boolean;
  confirmButtonText?: string;
  cancelButtonText?: string;
  confirmCallback?: () => void;
  cancelCallback?: () => void;
}

/**
 * Modal to warn a user about something related to an action.
 */
export const ConfirmAlertDialog = React.forwardRef(function (
  props: ConfirmAlertDialogProps,
  ref: React.Ref<HTMLElement>,
): React.ReactElement {
  const {
    open,
    title,
    content,
    showIcon,
    confirmButtonText,
    cancelButtonText,
    confirmCallback,
    cancelCallback,
    keepMounted,
  } = props;

  const handleCancelButton = () => {
    cancelCallback && cancelCallback();
  };

  const handleConfirmButton = () => {
    confirmCallback && confirmCallback();
  };
  const classes = useStyles();

  return (
    <div>
      <Dialog
        ref={ref}
        open={open}
        TransitionComponent={Transition}
        keepMounted={keepMounted}
        aria-labelledby="alert-dialog-slide-title"
        aria-describedby="alert-dialog-slide-description"
      >
        {showIcon && (
          <div className={classes.icon}>
            <WarningIcon
              color={'secondary'}
              style={{
                fontSize: '8rem',
              }}
            />
          </div>
        )}

        <DialogTitle id="alert-dialog-slide-title" className={classes.title}>
          {title || 'Are you sure you want to continue?'}
        </DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-slide-description">
            {content || 'Once you accept this there is no turning back.'}
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleCancelButton} color="primary" id="alert-dialog-cancel-button">
            {cancelButtonText || 'No'}
          </Button>
          <Button onClick={handleConfirmButton} color="secondary" id="alert-dialog-confirm-button">
            {confirmButtonText || `Yes, I'm sure.`}
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
});

const useStyles = makeStyles(() => ({
  icon: {
    height: '8rem',
    alignSelf: 'center',
  },
  title: {
    alignSelf: 'center',
  },
}));
