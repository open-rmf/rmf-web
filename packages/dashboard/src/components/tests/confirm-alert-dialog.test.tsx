import { mount, shallow } from 'enzyme';
import React from 'react';
import { ConfirmAlertDialog } from '../confirm-alert-dialog';

describe('Verification Alert', () => {
  test('Correct generation of default params values', () => {
    const root = shallow(<ConfirmAlertDialog open={true}></ConfirmAlertDialog>);
    expect(root).toMatchSnapshot();
    root.unmount();
  });

  test('Correct render with custom params values', () => {
    const root = shallow(
      <ConfirmAlertDialog
        open={true}
        title={'test'}
        content={'test'}
        confirmButtonText={'test'}
        cancelButtonText={'test'}
        cancelCallback={() => jest.mock}
        confirmCallback={() => jest.mock}
      ></ConfirmAlertDialog>,
    );
    expect(root).toMatchSnapshot();
    root.unmount();
  });

  test('Cancel callback is executed', () => {
    const cancelCallback = jest.fn();
    const root = mount(
      <ConfirmAlertDialog
        open={true}
        title={'test'}
        content={'test'}
        confirmButtonText={'test'}
        cancelButtonText={'test'}
        cancelCallback={cancelCallback}
        confirmCallback={() => jest.mock}
      ></ConfirmAlertDialog>,
    );
    root.find('#alert-dialog-cancel-button').first().simulate('click');
    expect(cancelCallback).toHaveBeenCalledTimes(1);
    root.unmount();
  });

  test('Confirm callback is executed', () => {
    const confirmCallback = jest.fn();
    const root = mount(
      <ConfirmAlertDialog
        open={true}
        title={'test'}
        content={'test'}
        confirmButtonText={'test'}
        cancelButtonText={'test'}
        cancelCallback={() => jest.mock}
        confirmCallback={confirmCallback}
      ></ConfirmAlertDialog>,
    );
    root.find('#alert-dialog-confirm-button').first().simulate('click');
    expect(confirmCallback).toHaveBeenCalledTimes(1);
    root.unmount();
  });

  test('Modal is closed', () => {
    let open = true;
    const closeModal = () => {
      open = false;
    };
    const root = mount(
      <ConfirmAlertDialog
        open={open}
        title={'test'}
        content={'test'}
        confirmButtonText={'test'}
        cancelButtonText={'test'}
        cancelCallback={() => jest.mock}
        confirmCallback={closeModal}
      ></ConfirmAlertDialog>,
    );
    root.find('#alert-dialog-confirm-button').first().simulate('click');
    expect(open).toBeFalsy();
    root.unmount();
  });
});
