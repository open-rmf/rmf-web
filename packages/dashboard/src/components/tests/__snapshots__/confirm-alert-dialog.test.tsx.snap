// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verification Alert Correct generation of default params values 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    TransitionComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    aria-describedby="alert-dialog-slide-description"
    aria-labelledby="alert-dialog-slide-title"
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      className="makeStyles-title-2"
      id="alert-dialog-slide-title"
    >
      Are you sure you want to continue?
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText))
        id="alert-dialog-slide-description"
      >
        Once you accept this there is no turning back.
      </WithStyles(ForwardRef(DialogContentText))>
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        id="alert-dialog-cancel-button"
        onClick={[Function]}
      >
        No
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="secondary"
        id="alert-dialog-confirm-button"
        onClick={[Function]}
      >
        Yes, I'm sure.
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;

exports[`Verification Alert Correct render with custom params values 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    TransitionComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    aria-describedby="alert-dialog-slide-description"
    aria-labelledby="alert-dialog-slide-title"
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      className="makeStyles-title-2"
      id="alert-dialog-slide-title"
    >
      test
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText))
        id="alert-dialog-slide-description"
      >
        test
      </WithStyles(ForwardRef(DialogContentText))>
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        id="alert-dialog-cancel-button"
        onClick={[Function]}
      >
        test
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="secondary"
        id="alert-dialog-confirm-button"
        onClick={[Function]}
      >
        test
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;
